proj3_chess/chess_pieces.py

```diff
@@ -0,0 +1,130 @@
+
+    def captures_piece(self, spot: tuple) -> bool:
+        """Recieves a spot on the board, and checks if it is occupied by an opponent's piece."""
+        spot_state = self.board.board_state[spot[0]][spot[1]]
+        return (
+            False
+            if (spot_state == None) or (spot_state.player == self.player)
+            else True
+        )
+
+    def available_moves(self) -> list:
+        """Returns a list of all moves available for the piece."""
+        print("'available_moves()' should be overritten by child classes")
+
+
+class Rook(ChessPiece):
+    piece_name = "R"
+
+    def available_moves(self) -> list:
+        moves = []
+        for x in (1, -1):
+            s = (self.spot[0] + x, self.spot[1])
+            while self.board.within_range(s) and self.legal_move(s):
+                moves.append((s, self.captures_piece(s)))
+                if self.captures_piece(s):
+                    break
+                s = (s[0] + x, s[1])
+        for y in (1, -1):
+            s = (self.spot[0], self.spot[1] + y)
+            while self.board.within_range(s) and self.legal_move(s):
+                moves.append((s, self.captures_piece(s)))
+                if self.captures_piece(s):
+                    break
+                s = (s[0], s[1] + y)
+        return moves
+
+
+class Bishop(ChessPiece):
+    piece_name = "B"
+
+    def available_moves(self) -> list:
+        moves = []
+        for x in (1, -1):
+            for y in (1, -1):
+                s = (self.spot[0] + x, self.spot[1] + y)
+                while self.board.within_range(s) and self.legal_move(s):
+                    moves.append((s, self.captures_piece(s)))
+                    if self.captures_piece(s):
+                        break
+                    s = (s[0] + x, s[1] + y)
+        return moves
+
+
+class Queen(Rook, Bishop):
+    piece_name = "Q"
+
+    def available_moves(self) -> list:
+        return Bishop.available_moves(self) + Rook.available_moves(self)
+
+
+class King(ChessPiece):
+    piece_name = "K"
+
+    def available_moves(self) -> list:
+        moves = []
+        for x in (-1, 0, 1):
+            for y in (-1, 0, 1):
+                s = (self.spot[0] + x, self.spot[1] + y)
+                if self.board.within_range(s) and s != self.spot and self.legal_move(s):
+                    moves.append((s, self.captures_piece(s)))
+        return moves
+
+
+class Knight(ChessPiece):
+    piece_name = "N"
+
+    def available_moves(self) -> list:
+        moves = []
+        for x in (-1, 1):
+            for y in (-1, 1):
+                for m in ((1, 2), (2, 1)):
+                    s = (self.spot[0] + m[0] * x, self.spot[1] + m[1] * y)
+                    if self.board.within_range(s) and self.legal_move(s):
+                        moves.append((s, self.captures_piece(s)))
+        return moves
+
+
+class Pawn(ChessPiece):
+    piece_name = "P"
+
+    def available_moves(self) -> list:
+        moves = []
```

Comment:
I'll admit I didn't trake time to fully comprehend the logic of this method because it is incredibly hard to read. Naming is very important, and also extracting code into private methods, even just for the sake of giving them a readable name